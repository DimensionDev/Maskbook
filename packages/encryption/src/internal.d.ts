type Readwrite<T> = {
    -readonly [key in keyof T]: T[key]
}
