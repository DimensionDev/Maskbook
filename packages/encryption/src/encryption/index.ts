export * from './Decryption'
export * from './Encryption'
export * from './AppendEncryption'
