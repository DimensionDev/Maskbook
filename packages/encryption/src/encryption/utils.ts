import type { EC_Public_CryptoKey, EC_Private_CryptoKey } from '@masknet/shared-base'
import { EC_KeyCurveEnum } from '../payload'
import type { EncryptIO } from './Encryption'

/** @internal */
export type EphemeralKeys = Map<
    EC_KeyCurveEnum,
    readonly [EC_Public_CryptoKey, EC_Private_CryptoKey] | Promise<readonly [EC_Public_CryptoKey, EC_Private_CryptoKey]>
>
/** @internal */
export function createEphemeralKeysMap(io: Pick<EncryptIO, 'getRandomECKey'>) {
    const ephemeralKeys: EphemeralKeys = new Map()

    // get ephemeral keys, generate one if not found
    async function getEphemeralKey(curve: EC_KeyCurveEnum) {
        if (ephemeralKeys.has(curve)) return ephemeralKeys.get(curve)!
        ephemeralKeys.set(curve, generateEC_KeyPair(io, curve))
        return ephemeralKeys.get(curve)!
    }
    return { ephemeralKeys, getEphemeralKey }
}

/** @internal */
export function fillIV(io: Pick<EncryptIO, 'getRandomValues'>): Uint8Array {
    if (io.getRandomValues) return io.getRandomValues(new Uint8Array(16))
    return crypto.getRandomValues(new Uint8Array(16))
}

async function generateEC_KeyPair(io: Pick<EncryptIO, 'getRandomECKey'>, kind: EC_KeyCurveEnum) {
    if (io.getRandomECKey) return io.getRandomECKey(kind)
    const namedCurve: Record<EC_KeyCurveEnum, string> = {
        [EC_KeyCurveEnum.secp256p1]: 'P-256',
        [EC_KeyCurveEnum.secp256k1]: 'K-256',
    }
    const { privateKey, publicKey } = await crypto.subtle.generateKey(
        { name: 'ECDH', namedCurve: namedCurve[kind] },
        true,
        ['deriveKey'],
    )
    return [publicKey as EC_Public_CryptoKey, privateKey as EC_Private_CryptoKey] as const
}
