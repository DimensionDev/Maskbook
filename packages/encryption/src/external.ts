// This file is the entry for external usage.
export {
    encodePayload,
    parsePayload,
    EC_KeyCurveEnum,
    SocialNetworkEnum,
    SocialNetworkEnumToProfileDomain,
    type EC_Key,
    type Signature,
    type PayloadParseResult,
    type PayloadWellFormed,
    type SupportedPayloadVersions,
} from './payload/index.js'

export {
    decrypt,
    DecryptError,
    DecryptProgressKind,
    DecryptIntermediateProgressKind,
    DecryptErrorReasons,
    type DecryptOptions,
    type DecryptIO,
    type DecryptEphemeralECDH_PostKey,
    type DecryptStaticECDH_PostKey,
    type DecryptProgress,
    type DecryptIntermediateProgress,
    type DecryptReportedInfo,
    type DecryptSuccess,
} from './encryption/index.js'

export { socialNetworkEncoder, socialNetworkDecoder } from './social-network-encode-decode/index.js'

export { type TypedMessage, type Meta } from '../../typed-message/dist/base/index.js'
export * from '../../typed-message/dist/base/core/index.js'
export {
    ProfileIdentifier,
    PostIVIdentifier,
    CheckedError,
    OptionalResult,
    type AESCryptoKey,
    type EC_Public_CryptoKey,
    type EC_Private_CryptoKey,
    type EC_CryptoKey,
} from '@masknet/shared-base'

export * from 'ts-results-es'
