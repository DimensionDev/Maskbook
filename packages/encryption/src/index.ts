export {
    encodePayload,
    parsePayload,
    EC_KeyCurveEnum,
    SocialNetworkEnum,
    SocialNetworkEnumToProfileDomain,
    type EC_Key,
    type Signature,
    type PayloadParseResult,
    type PayloadWellFormed,
    type SupportedPayloadVersions,
} from './payload'

export {
    encrypt,
    EncryptError,
    EncryptErrorReasons,
    type EncryptOptions,
    type EncryptIO,
    type EncryptResult,
    type EncryptTargetE2E,
    type EncryptionResultE2EMap,
    type EncryptTargetPublic,
    type EncryptionResultE2E,
} from './encryption'

export {
    decrypt,
    DecryptError,
    DecryptProgressKind,
    DecryptIntermediateProgressKind,
    DecryptErrorReasons,
    type DecryptOptions,
    type DecryptIO,
    type DecryptEphemeralECDH_PostKey,
    type DecryptStaticECDH_PostKey,
    type DecryptProgress,
    type DecryptIntermediateProgress,
    type DecryptReportedInfo,
    type DecryptSuccess,
} from './encryption'

export {
    //
    type AppendEncryptionIO,
    type AppendEncryptionOptions,
    appendEncryptionTarget,
} from './encryption'

export { socialNetworkEncoder, socialNetworkDecoder, TwitterDecoder } from './social-network-encode-decode'

export {
    type DecodeImageOptions,
    type EncodeImageOptions,
    type ImageTemplateTypes,
    type SteganographyIO,
    steganographyDecodeImage,
    steganographyEncodeImage,
    GrayscaleAlgorithm,
} from './image-steganography'

// TODO: remove them in the future
export { importEC_Key, getEcKeyCurve, importAES as importAESFromJWK } from './utils'
