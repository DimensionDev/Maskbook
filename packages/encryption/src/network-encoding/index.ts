import { EncryptPayloadNetwork } from '../payload/types.js'
import { sharedDecoder, sharedEncoder } from './shared.js'
import { TwitterDecoder, __TwitterEncoder } from './twitter.js'

export function decodeByNetwork(network: EncryptPayloadNetwork, content: string): Array<string | Uint8Array> {
    if (network === EncryptPayloadNetwork.Twitter) {
        return TwitterDecoder(content)
            .map((x) => [x])
            .unwrapOr([])
    }
    const possiblePayload = content.match(/(\u{1F3BC}[\w+/=|]+:\|\|)/giu) || []

    const result: Array<string | Uint8Array> = []
    for (const payload of possiblePayload) {
        const decoded = sharedDecoder(payload)
        if (decoded.isSome()) result.push(decoded.value)
        else result.push(payload)
    }
    return result
}
export function encodeByNetwork(network: EncryptPayloadNetwork, content: string | Uint8Array): string {
    // v38
    if (typeof content === 'string') {
        if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
        return content
    }

    // v37
    if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
    return sharedEncoder(content)
}

export * from './twitter.js'
