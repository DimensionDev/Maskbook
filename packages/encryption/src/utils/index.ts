import { ExceptionKinds, Exception } from '../types'
import { Result } from 'ts-results'
import { encodeArrayBuffer, decodeArrayBuffer, encodeText, decodeText } from '@dimensiondev/kit'

export * from './crypto'

const firstArgString = (e: unknown) => typeof e !== 'string'
const firstArgArrayBuffer = (e: unknown) => e instanceof ArrayBuffer
export const encodeArrayBufferF = wrap(encodeArrayBuffer, firstArgArrayBuffer)
export const decodeArrayBufferF = wrap(decodeArrayBuffer as (x: string) => ArrayBuffer, firstArgString)
export const encodeTextF = wrap(encodeText, firstArgString)
export const decodeTextF = wrap(decodeText, firstArgArrayBuffer)
export const JSONParseF = wrap(JSON.parse, firstArgString)

function wrap<P extends any[], T>(f: (...args: P) => T, valid: (...args: P) => boolean) {
    return <E extends ExceptionKinds>(invalidE: E, throwsE: E) =>
        (...args: P) => {
            const isValid = valid(...args)
            if (!isValid) return new Exception(invalidE, null).toErr()
            return Result.wrap(() => f(...args)).mapErr(Exception.mapErr(throwsE))
        }
}
export async function andThenAsync<T, E, Q, E2>(
    op: Result<T, E> | Promise<Result<T, E>>,
    mapper: (t: T) => Result<Q, E2> | Promise<Result<Q, E2>>,
): Promise<Result<Q, E | E2>> {
    op = await op
    if (op.err) return op
    return mapper(op.val)
}
