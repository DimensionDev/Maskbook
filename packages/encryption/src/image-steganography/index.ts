import { encode, decode, GrayscaleAlgorithm } from '@dimensiondev/stego-js'
import type { EncodeOptions } from '@dimensiondev/stego-js'
import { getDimension } from './utils.js'
import { getPreset, findPreset } from './presets.js'
import { encodeArrayBuffer } from '@dimensiondev/kit'

export { GrayscaleAlgorithm, AlgorithmVersion } from '@dimensiondev/stego-js'

export interface SteganographyIO {
    downloadImage: (url: string) => Promise<ArrayBuffer>
}
export interface EncodeImageOptions extends SteganographyIO {
    data: string | Uint8Array
    password: string
    grayscaleAlgorithm?: GrayscaleAlgorithm
    preset: SteganographyPreset
}
export enum SteganographyPreset {
    Preset2021 = '2021',
    Preset2022 = '2022',
}

export async function steganographyEncodeImage(buf: ArrayBuffer, options: EncodeImageOptions) {
    const { downloadImage, data, password, grayscaleAlgorithm } = options
    const preset = getPreset(options.preset)
    if (!preset) throw new Error('Failed to find preset.')

    const optionalOptions: Partial<Readwrite<EncodeOptions>> = {}
    if (grayscaleAlgorithm) optionalOptions.grayscaleAlgorithm = grayscaleAlgorithm

    const text = typeof data === 'string' ? data : encodeArrayBuffer(data)

    return new Uint8Array(
        await encode(buf, await downloadImage(preset.mask), {
            ...preset.options,
            ...optionalOptions,
            text,
            pass: password,
        }),
    )
}

export type DecodeImageOptions = SteganographyIO & Partial<EncodeOptions> & Pick<EncodeOptions, 'pass'>

export async function steganographyDecodeImage(image: Blob | string, options: DecodeImageOptions) {
    const buffer = typeof image === 'string' ? await options.downloadImage(image) : await image.arrayBuffer()
    const dimension = getDimension(buffer)
    const preset = findPreset(dimension)
    if (!preset) return ''
    return decode(buffer, await options.downloadImage(preset.mask), {
        ...preset.options,
        ...options,
    })
}
