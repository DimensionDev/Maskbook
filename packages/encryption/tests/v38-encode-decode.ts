import { expect, test } from '@jest/globals'
import { encodePayload, parsePayload, PayloadWellFormed } from '../src'
import { None, Some } from 'ts-results'
import { ProfileIdentifier } from '@masknet/shared-base'
import { importAES } from '../src/utils'
import { queryTestPublicKey } from './keys'

test('Parse v38 encoded by old infra', async () => {
    const out = (await parsePayload(oldInfraOutput)).unwrap()
    expect(out).toMatchSnapshot('Parse v38 payload result')
})

test('Parse older v38 payload that does not have newer field', async () => {
    const out = (await parsePayload(oldInfraOutputShort)).unwrap()
    expect(out).toMatchSnapshot('Parse older v38 payload result')
})

test('Encode v38 payload', async () => {
    const payload: PayloadWellFormed.Payload = {
        author: Some(new ProfileIdentifier('facebook.com', 'test')),
        authorPublicKey: Some((await queryTestPublicKey(new ProfileIdentifier('localhost', 'alice')))!),
        encrypted: new Uint8Array(Buffer.from('3a0d6ee692c6f46896b196f14301c01ad2fa26aa', 'hex')),
        encryption: {
            type: 'public',
            iv: new Uint8Array(Buffer.from('0633db7e24805c2bdcff69ea2afda7cd', 'hex')),
            AESKey: await importAES(AESKey).then((x) => x.unwrap()),
        },
        signature: None,
        version: -38,
    }

    const result = await encodePayload.NoSign(payload).then((x) => x.unwrap())
    expect(result).toMatchSnapshot('Encoded v38 message')

    const parsed = await parsePayload(result)
    expect(parsed).toMatchSnapshot('Parse generated v38 message')
})

/* cspell:disable */
const oldInfraOutput =
    '\u{1F3BC}4/4|avkwBKqMpCKznGclvChuuh2AEExV0J14xI/KANhwiKJfVyfm2ObWb432E3aAOa7ImRoCd7/JK1d' +
    'DQWk4rt9NqajTEaajARMc9hJ9GmR8lorBNRNHlgj/h1KJYk5th7Nsr04PWO0nJUKiDH2CJwieSxW2YqxC' +
    'I1ceYKUYcZOsVJEZOrJ/IB8WUmU0|BjPbfiSAXCvc/2nqKv2nzQ==|Og1u5pLG9GiWsZbxQwHAGtL6Jqo' +
    '=|_|Aq/bVWAKvodJuURGk3enjE1gUiu2SELM8IIKIlNGqOWM|1|ZmFjZWJvb2suY29tLzEwMDAyNzU2MjI' +
    '0OTU3NA==:||'
const oldInfraOutputShort =
    '\u{1F3BC}4/4|avkwBKqMpCKznGclvChuuh2AEExV0J14xI/KANhwiKJfVyfm2ObWb432E3aAOa7ImRoCd7/JK1d' +
    'DQWk4rt9NqajTEaajARMc9hJ9GmR8lorBNRNHlgj/h1KJYk5th7Nsr04PWO0nJUKiDH2CJwieSxW2YqxC' +
    'I1ceYKUYcZOsVJEZOrJ/IB8WUmU0|BjPbfiSAXCvc/2nqKv2nzQ==|Og1u5pLG9GiWsZbxQwHAGtL6Jqo=|_:||'
/* cspell:enable */

const AESKey = {
    key_ops: ['encrypt', 'decrypt'],
    ext: true,
    kty: 'oct',
    /* cspell:disable-next-line */
    k: 'JrotLWI_e9OUOXzONFPthyMq-EyHdtp9vlAE9iAI9Gc',
    alg: 'A256GCM',
}
