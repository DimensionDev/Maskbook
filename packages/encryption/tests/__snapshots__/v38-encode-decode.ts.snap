// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encode v38 payload: Encoded v38 message 1`] = `"ðŸŽ¼4/4|avkwBKqMpCKznGclvChuuh2AEExV0J14xI/KANhwiKJfVyfm2ObWb432E3aAOa7ImRoCd7/JK1dDQWk4rt9NqajTEaajARMc9hJ9GmR8lorBNRNHlgj/h1KJYk5th7Nsr04PWO0nJUKiDH2CJwieSxW2YqxCI1ceYKUYcZOsVJEZOrJ/IB8WUmU0|BjPbfiSAXCvc/2nqKv2nzQ==|Og1u5pLG9GiWsZbxQwHAGtL6Jqo=|_|Aq/bVWAKvodJuURGk3enjE1gUiu2SELM8IIKIlNGqOWM|1|ZmFjZWJvb2suY29tL3Rlc3Q=:||"`;

exports[`Encode v38 payload: Parse generated v38 message 1`] = `
Ok(Object {
  "author": Ok(Some(ProfileIdentifier {
    "network": "facebook.com",
    "userId": "test",
  })),
  "authorPublicKey": Ok(Some(Object {
    "algr": 2,
    "key": CryptoKey { [opaque crypto key material] },
  })),
  "encrypted": Ok(Uint8Array [ 3a0d6ee692c6f46896b196f14301c01ad2fa26aa ]),
  "encryption": Ok(Object {
    "AESKey": Ok(CryptoKey { [opaque crypto key material] }),
    "iv": Ok(Uint8Array [ 0633db7e24805c2bdcff69ea2afda7cd ]),
    "type": "public",
  }),
  "signature": Ok(None),
  "version": -38,
})
`;

exports[`Parse older v38 payload that does not have newer field: Parse older v38 payload result 1`] = `
Object {
  "author": Ok(None),
  "authorPublicKey": Ok(None),
  "encrypted": Ok(Uint8Array [ 3a0d6ee692c6f46896b196f14301c01ad2fa26aa ]),
  "encryption": Ok(Object {
    "ephemeralPublicKey": Object {},
    "iv": Ok(Uint8Array [ 0633db7e24805c2bdcff69ea2afda7cd ]),
    "ownersAESKeyEncrypted": Ok(Uint8Array [ 6af93004aa8ca422b39c6725bc286eba1d80104c55d09d78c48fca00d87088a25f5727e6d8e6d66f8df613768039aec8991a0277bfc92b5743416938aedf4da9a8d311a6a301131cf6127d1a647c968ac13513479608ff875289624e6d87b36caf4e0f58ed272542a20c7d8227089e4b15b662ac4223571e60a5187193ac5491193ab27f201f16526534 ]),
    "type": "E2E",
  }),
  "signature": Ok(None),
  "version": -38,
}
`;

exports[`Parse v38 encoded by old infra: Parse v38 payload result 1`] = `
Object {
  "author": Ok(Some(ProfileIdentifier {
    "network": "facebook.com",
    "userId": "100027562249574",
  })),
  "authorPublicKey": Ok(Some(Object {
    "algr": 2,
    "key": CryptoKey { [opaque crypto key material] },
  })),
  "encrypted": Ok(Uint8Array [ 3a0d6ee692c6f46896b196f14301c01ad2fa26aa ]),
  "encryption": Ok(Object {
    "AESKey": Ok(CryptoKey { [opaque crypto key material] }),
    "iv": Ok(Uint8Array [ 0633db7e24805c2bdcff69ea2afda7cd ]),
    "type": "public",
  }),
  "signature": Ok(None),
  "version": -38,
}
`;
